### ISSUE TITLE
Periodically seeing Errno::EPIPE "Broken pipe" errors on SSH Linux targets

### ISSUE THREAD

**Issue by nmaludy** [puppetlabs/bolt#1769]
## Describe the Bug
Periodically, when targetting Linux hosts with SSH i'm getting the following error on a task: 
```json
2020-04-27T07:25:40.005535 INFO   Bolt::Executor: {"target":"hostname.domain.tld","action":"task","object":null,"status":"failure","value":{"_error":{"kind":"puppetlabs.tasks/exception-error","issue_code":"EXCEPTION","msg":"Broken pipe","details":{"class":"Errno::EPIPE","stack_trace":"/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:153:in `write'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:153:in `handle_sudo'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:169:in `block in check_sudo'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:169:in `map!'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:169:in `check_sudo'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:354:in `execute'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:125:in `block (2 levels) in run_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:263:in `with_tempdir'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:89:in `block in run_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:224:in `running_as'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:77:in `run_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/simple.rb:43:in `block in run_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/ssh.rb:27:in `with_connection'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/simple.rb:42:in `run_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/base.rb:108:in `block in batch_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/base.rb:50:in `with_events'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/base.rb:106:in `batch_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/executor.rb:275:in `block (3 levels) in run_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/executor.rb:236:in `with_node_logging'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/executor.rb:274:in `block (2 levels) in run_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/executor.rb:115:in `block (3 levels) in queue_execute'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:348:in `run_task'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:337:in `block (3 levels) in create_worker'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `loop'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `block (2 levels) in create_worker'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `catch'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `block in create_worker'\\n/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'"}}}}
```

I extracted the stacktrace and unescaped the newlines which resulted in this:
```shell
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:153:in `write'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:153:in `handle_sudo'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:169:in `block in check_sudo'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:169:in `map!'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:169:in `check_sudo'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:354:in `execute'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:125:in `block (2 levels) in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:263:in `with_tempdir'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:89:in `block in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:224:in `running_as'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/shell/bash.rb:77:in `run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/simple.rb:43:in `block in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/ssh.rb:27:in `with_connection'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/simple.rb:42:in `run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/base.rb:108:in `block in batch_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/base.rb:50:in `with_events'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/transport/base.rb:106:in `batch_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/executor.rb:275:in `block (3 levels) in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/executor.rb:236:in `with_node_logging'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/executor.rb:274:in `block (2 levels) in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.6.0/lib/bolt/executor.rb:115:in `block (3 levels) in queue_execute'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:348:in `run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:337:in `block (3 levels) in create_worker'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `loop'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `block (2 levels) in create_worker'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `catch'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `block in create_worker'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

I'm not sure why this is happening. I haven't been able to come up with a set of reproduction steps yet. The commonalities between the failures are:
- It always happens on a Linux host (observed on CentOS 7 and RHEL 7, we don't have much Ubuntu/Debian at all so probably not ever going to see i there)
- It is always a task that is executing a bash script
- It appears to always fails in the `sudo` part of the code

Other observations:
- Fails on different hosts each time, across different customer environments
- Fails on different tasks each time, so i don't think it's a task bug
- Failed across several version of Bolt, seen on both `2.5.0` and `2.6.0`

Stack trace on `2.6.0` is above. Stack Trace on `2.5.0` is below:
```
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:153:in `write'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:153:in `handle_sudo'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:169:in `block in check_sudo'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:169:in `map!'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:169:in `check_sudo'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:354:in `execute'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:125:in `block (2 levels) in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:263:in `with_tempdir'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:89:in `block in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:224:in `running_as'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/shell/bash.rb:77:in `run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/transport/simple.rb:43:in `block in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/transport/ssh.rb:27:in `with_connection'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/transport/simple.rb:42:in `run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/transport/base.rb:108:in `block in batch_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/transport/base.rb:50:in `with_events'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/transport/base.rb:106:in `batch_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/executor.rb:276:in `block (3 levels) in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/executor.rb:237:in `with_node_logging'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/executor.rb:275:in `block (2 levels) in run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/bolt-2.5.0/lib/bolt/executor.rb:116:in `block (3 levels) in queue_execute'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:348:in `run_task'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:337:in `block (3 levels) in create_worker'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `loop'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `block (2 levels) in create_worker'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `catch'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `block in create_worker'
/opt/puppetlabs/bolt/lib/ruby/gems/2.5.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

## Expected Behavior
No exceptions are thrown.

## Steps to Reproduce
Unable to successfully reproduce yet

## Environment
 - Bolt `2.5.0` and `2.6.0`
 - Platform `CentOS 7` and `RHEL 7`

**Comment by nicklewis** [comment]
It looks like this will happen if the task exits before Bolt can finish writing all of STDIN (so presumably the task doesn't actually _use_ STDIN in this case). This bug appears to have already 
been present with the `local` transport when using `run-as` and was ported to the SSH transport when that code was generalized in #1750 

We actually explicitly handle this case (even with a comment explaining what can happen) when _not_ using sudo, but don't handle it for sudo. I'll get together a fix. Thanks for reporting this.

**Comment by nicklewis** [comment]
There's a related issue here that causes a deadlock in the case where a) both stdin and stdout are larger than the pipe buffer, b) you're using run-as, c) the task doesn't read stdin. The task will block trying to write stdout, but Bolt is blocked trying to write stdin.

When not using run-as, Bolt will do the right thing and `select()` on all the streams, but for run-as it just tries to write stdin in one go. That bug will require a bit more of a rework to the way we handle sudo, so I might deal with it separately and just get up the PR for this.

**Comment by nicklewis** [comment]
Closed by #1775
